In [10]:
import pandas as pd
import regex as re

# must be specified to be able to use methods of package
import sys
sys.path.append('../src') 
# import packages

from src.political_ads.Preprocessor import *
from src.political_ads.keyword_filter import *

### Comparing Zeroshot with the Keyword search approach


Below are the preprocessing methods applied for both data samples

The following section cleans, removes dubs from each sample and then compares the two: keyword vs. zeroshot

In [11]:
# preprocessing methods appplied to both samples

# Remove links
def clean_data(dataframe):
#replace URL of a text
    dataframe['ad_creative_body'] = dataframe['ad_creative_body'].str.replace('http[s]?://(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*\(\),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+', ' ')



# Remove emojis characters
def remove_emojis(data):
    try:
      emoj = re.compile("["
          u"\U0001F600-\U0001F64F"  # emoticons
          u"\U0001F300-\U0001F5FF"  # symbols & pictographs
          u"\U0001F680-\U0001F6FF"  # transport & map symbols
          u"\U0001F1E0-\U0001F1FF"  # flags (iOS)
          u"\U00002500-\U00002BEF"  # chinese char
          u"\U00002702-\U000027B0"
          u"\U00002702-\U000027B0"
          u"\U000024C2-\U0001F251"
          u"\U0001f926-\U0001f937"
          u"\U00010000-\U0010ffff"
          u"\u2640-\u2642" 
          u"\u2600-\u2B55"
          u"\u200d"
          u"\u23cf"
          u"\u23e9"
          u"\u231a"
          u"\ufe0f"  # dingbats
          u"\u3030"
                        "]+", re.UNICODE)
      return re.sub(emoj, '', data)
    except:
      return data

def remove_punctuations(text):
    return re.sub('[^A-Za-z0-9]+', ' ', text)
  
  #Consider stemming and lemmatizing
def preprocess_df(input: pd.DataFrame) -> pd.DataFrame:
  """makes the df ready by lowercasing, removing emojis and numbers"""
  output = input.copy()
  # Lowercase the text
  output["ad_creative_body"] = output["ad_creative_body"].str.lower()

  # Apply emoji removal to df
  output["ad_creative_body"] = output["ad_creative_body"].map(remove_emojis)

  output["ad_creative_body"] = output["ad_creative_body"].map(remove_punctuations)
 
  # Remove all numbers from text
  #output['Title'] = output['Title'].str.replace('\d+', '')
  #output['Description'] = output['Description'].str.replace('\d+', '')

  return output


In [12]:
#Code to read from our full dataset:
preprocess = Preprocessor()
df = preprocess.file_to_df('/home/gustavgyrst/Desktop/Research Project/political-ad-api/data/all_politicians_aggregated.txt')
df.head(3)


ad_creation_time                                   ad_creative_body   spend  \
0       2021-10-14  To show that you're part of our fight to prote...  1749.5   
1       2021-10-14  Nothing is spookier than the ongoing assault o...   149.5   
2       2021-10-14  To show that you're part of our fight to prote...  1749.5   

   impressions                                 delivery_by_region  \
0      74999.5  [{'percentage': '0.004793', 'region': 'Alabama...   
1      27499.5  [{'percentage': '0.005051', 'region': 'Alabama...   
2     162499.5  [{'percentage': '0.004873', 'region': 'Alabama...   

                            demographic_distribution      page_id  \
0  [{'percentage': '0.010689', 'age': '18-24', 'g...  38471053686   
1  [{'percentage': '0.000629', 'age': '18-24', 'g...  38471053686   
2  [{'percentage': '0.005081', 'age': '18-24', 'g...  38471053686   

          page_name                 bylines               id  spend_lo  \
0  Elizabeth Warren  WARREN DEMOCRATS, INC.  195916456012328      1500   
1  Elizabeth Warren  WARREN DEMOCRATS, INC.  197833945796073       100   
2  Elizabeth Warren  WARREN DEMOCRATS, INC.  396978045241497      1500   

   spend_hi  impressions_lo  impressions_hi  
0      1999           70000           79999  
1       199           25000           29999  
2      1999          150000          174999

In [13]:
#getting the climate ads from the total raw dataset
c_filter = Filter()
df_climate = c_filter.get_climate_ads(df)


In [14]:
#df_climate_nodub = df_climate[df_climate['ad_creative_body'].duplicated()==False]

# Another approach / same result
df_climate_nodub = df_climate.drop_duplicates(subset='ad_creative_body', keep='last')

In [15]:
df_climate_nodub

ad_creation_time                                   ad_creative_body  \
4774         2020-02-10  Every kid in America should have the same acce...   
6458         2020-01-26  I’m proposing a tax on the top 0.1% — the Jeff...   
8230         2020-01-06  Elizabeth Warren is returning to New Hampshire...   
9682         2019-12-20  Climate scientists published new research sugg...   
9683         2019-12-20  Climate scientists published new research sugg...   
...                 ...                                                ...   
602149       2019-05-06  Climate change is one of the most pressing iss...   
602154       2019-05-06  Climate change is one of the most pressing iss...   
602157       2019-05-06  Climate change is one of the most pressing iss...   
602162       2019-05-06  Climate change is one of the most pressing iss...   
602511       2020-09-15  We deserve better in the 170th. We deserve som...   

        spend  impressions                                 delivery_by_region  \
4774     49.5        499.5  [{'percentage': '0.007273', 'region': 'Alabama...   
6458     49.5        499.5                                                NaN   
8230    549.5      32499.5  [{'percentage': '0.001361', 'region': 'Connect...   
9682    249.5       9499.5  [{'percentage': '0.006624', 'region': 'Rhode I...   
9683     49.5       1499.5  [{'percentage': '0.000701', 'region': 'Delawar...   
...       ...          ...                                                ...   
602149   49.5        499.5                                                NaN   
602154   49.5        499.5        [{'percentage': '1', 'region': 'Colorado'}]   
602157   49.5        499.5                                                NaN   
602162   49.5        499.5        [{'percentage': '1', 'region': 'Colorado'}]   
602511  149.5      12499.5    [{'percentage': '1', 'region': 'Pennsylvania'}]   

                                 demographic_distribution          page_id  \
4774    [{'percentage': '0.407273', 'age': '25-34', 'g...      38471053686   
6458                                                  NaN      38471053686   
8230    [{'percentage': '0.107619', 'age': '45-54', 'g...      38471053686   
9682    [{'percentage': '0.006735', 'age': '25-34', 'g...      38471053686   
9683    [{'percentage': '0.000708', 'age': '65+', 'gen...      38471053686   
...                                                   ...              ...   
602149                                                NaN  280704452367167   
602154  [{'percentage': '0.090278', 'age': '45-54', 'g...  280704452367167   
602157                                                NaN  280704452367167   
602162  [{'percentage': '0.026087', 'age': '45-54', 'g...  280704452367167   
602511  [{'percentage': '0.001199', 'age': '65+', 'gen...  174750213023790   

                                               page_name  \
4774                                    Elizabeth Warren   
6458                                    Elizabeth Warren   
8230                                    Elizabeth Warren   
9682                                    Elizabeth Warren   
9683                                    Elizabeth Warren   
...                                                  ...   
602149                                        Jason Crow   
602154                           Jason Crow for Congress   
602157                           Jason Crow for Congress   
602162                                        Jason Crow   
602511  Mike Doyle for PA Representative, 170th District   

                            bylines                id  spend_lo  spend_hi  \
4774     WARREN FOR PRESIDENT, INC.  1044291342615870         0        99   
6458     WARREN FOR PRESIDENT, INC.   193650885156389         0        99   
8230     WARREN FOR PRESIDENT, INC.  1026661577694555       500       599   
9682     WARREN FOR PRESIDENT, INC.   580884722482084       200       299   
9683     WARREN FOR PRESIDENT, INC.  2750300318326852 

In [16]:
# apply this to the dataframe
clean_data(df_climate_nodub)

# apply it to df
keywords_cleaned = preprocess_df(df_climate_nodub)

/tmp/ipykernel_19229/1943526736.py:6: FutureWarning: The default value of regex will change from True to False in a future version.
  dataframe['ad_creative_body'] = dataframe['ad_creative_body'].str.replace('http[s]?://(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*\(\),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+', ' ')
/tmp/ipykernel_19229/1943526736.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataframe['ad_creative_body'] = dataframe['ad_creative_body'].str.replace('http[s]?://(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*\(\),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+', ' ')


In [17]:
#Adding label for use when merging the two datasets later down
keywords_cleaned['sample'] = 'climate_keywords'

In [18]:
keywords_cleaned = keywords_cleaned[['ad_creative_body', 'sample']]

In [19]:
keywords_cleaned

ad_creative_body            sample
4774    every kid in america should have the same acce...  climate_keywords
6458    i m proposing a tax on the top 0 1 the jeff be...  climate_keywords
8230    elizabeth warren is returning to new hampshire...  climate_keywords
9682    climate scientists published new research sugg...  climate_keywords
9683    climate scientists published new research sugg...  climate_keywords
...                                                   ...               ...
602149  climate change is one of the most pressing iss...  climate_keywords
602154  climate change is one of the most pressing iss...  climate_keywords
602157  climate change is one of the most pressing iss...  climate_keywords
602162  climate change is one of the most pressing iss...  climate_keywords
602511  we deserve better in the 170th we deserve some...  climate_keywords

[1339 rows x 2 columns]

Same steps are applied to the zeroshot sample...

In [20]:
#zeroshot
zeroshot = pd.read_csv("unique_ads_preprocessed_labeled.csv")

In [21]:
#example of a dub
dub = zeroshot[zeroshot.ad_creative_body == 'conservatives it s time to act the attacks from the left are relentless but we don t have to standby and take the abuse we can unite with more than 2 3 million fellow conservatives to preserve and protect our great nation by joining amac amac the association of mature american citizens is the conservative alternative to aarp the fight for our freedoms and our cherished american values has never been more intense the great folks at amac believe america is a great country because america is a good country and that those three words that start our constitution we the people still mean something still matter so do the 2 3 million members who count on amac to fight for the things they care about their freedom their values amac also works hard to deliver real value to their members providing the best benefits discounts and services found in one place stand up and show your support for our great country by joining amac www amac us carter as you know we are up against a formidable army of liberal companies media celebrities lobbyists activists and politicians if we keep spending our hard earned money with aarp or do nothing then we cannot expect to win this fight we can counter the left if we come together on issues that are important to us a united voice is a strong voice and with your help we can move the needle back to the right being a part of a great organization like amac gives us the opportunity to make a difference amac offers 4 great membership options each one includes you and your spouse here are just a few of the member benefits waiting for you medicare insurance plans social security advisor under 65 health insurance prescription discount card discount dental and vision plans auto and home insurance roadside assistance discounts travel lodging rental cars restaurants financial planning join the fight join amac today click or call 844 887 2622']

In [22]:
#Dropping dubs - approxx 7000
zeroshot = zeroshot.drop_duplicates(subset='ad_creative_body', keep='last')

In [23]:
zeroshot

ad_creative_body  \
0      it is not enough to defeat donald trump. we mu...   
1      we just won a huge victory in the primary elec...   
2      i’ve got a small favor to ask – can you add yo...   
3      the dnc just raised the debate donor threshold...   
4      the enormous student debt crisis is the result...   
...                                                  ...   
57842  nancy pelosi's leading the charge to violate t...   
57843  the consequences of war is that our daughters ...   
57844  nancy pelosi's impeachment hit-list keeps grow...   
57845  nancy pelosi's agenda for congress does not re...   
57846  the supreme court just started hearings on the...   

                                            impr_geo_agg  spend_agg  \
0      {'Illinois': 2262508, 'Ohio': 2400000, 'Texas'...   803277.5   
1      {'South Dakota': 40245, 'South Carolina': 1640...   349932.5   
2      {'Alaska': 58937, 'Pennsylvania': 692570, 'Nev...   271561.5   
3      {'Alabama': 139291, 'Pennsylvania': 799612, 'N...   616753.5   
4      {'Alabama': 324699, 'Pennsylvania': 748758, 'N...   221751.5   
...                                                  ...        ...   
57842                                   {'Florida': 500}       49.5   
57843                                     {'Maine': 500}       49.5   
57844                                  {'Nebraska': 500}       49.5   
57845                                     {'Texas': 500}       49.5   
57846                                  {'Delaware': 500}       49.5   

         impr_agg           page_id                 page_name  \
0      25442483.5   124955570892789            Bernie Sanders   
1      24854683.0  1316372698453411  Alexandria Ocasio-Cortez   
2      19366864.5   124955570892789            Bernie Sanders   
3      18443203.5       36872302227               Cory Booker   
4      18399401.5       38471053686          Elizabeth Warren   
...           ...               ...                       ...   
57842       499.5   666196620070127      Maria Elvira Salazar   
57843       499.5   344010372303511                Angus King   
57844       499.5   109592402468562               Deb Fischer   
57845       499.5  1558042497785635            Beth Van Duyne   
57846       499.5        7209432197               Chris Coons   

      ad_creation_time        party  \
0           2020-03-05     Democrat   
1           2020-09-11     Democrat   
2           2020-03-17     Democrat   
3           2019-07-25     Democrat   
4           2020-02-04     Democrat   
...                ...          ...   
57842       2021-10-21   Republican   
57843       2018-08-07  Independent   
57844       2019-11-11   Republican   
57845       2020-04-20   Republican   
57846       2020-04-23     Democrat   

                                                emotions     ANGRY   ANNOYED  \
0      {'AFRAID': 0.09056519009021387, 'AMUSED': 0.16...  0.116733  0.135540   
1      {'AFRAID': 0.09745483939705116, 'AMUSED': 0.13...  0.134677  0.142094   
2      {'AFRAID': 0.07969887694789736, 'AMUSED': 0.15...  0.131562  0.158462   
3      {'AFRAID': 0.09375970552559433, 'AMUSED': 0.13...  0.133796  0.140846   
4      {'AFRAID': 0.10524118111376297, 'AMUSED': 0.12...  0.139728  0.137710   
...                                                  ...       ...       ...   
57842  {'AFRAID': 0.11236538195107602, 'AMUSED': 0.13...  0.168018  0.161798   
57843  {'AFRAID': 0.09867305237088528, 'AMUSED': 0.13...  0.125781  0.135670   
57844  {'AFRAID': 0.0816414466029903, 'AMUSED': 0.140...  0.141876  0.180827   
57845  {'AFRAID': 0.0887159273868603, 'AMUSED': 0.144...  0.129105  0.155833   
57846  {'AFRAID': 0.10437548498299842, 'AMUSED': 0.12...  0.168615  0.157040   

         AMUSED    AFRAID     HAPPY  INSPIRED       SAD  DONT_CARE  \
0      0.166387  0.090565  0.097104  0.184500  0.088420   0.120751   
1      0.130600  0.097455  0.103810  0.171917  0.099821   0.119625   
2      0.151987  0.079699  0.108307  0.165868

In [24]:
# apply this to the dataframe
clean_data(zeroshot)

# apply it to df
zeroshot_cleaned = preprocess_df(zeroshot)

/tmp/ipykernel_19229/1943526736.py:6: FutureWarning: The default value of regex will change from True to False in a future version.
  dataframe['ad_creative_body'] = dataframe['ad_creative_body'].str.replace('http[s]?://(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*\(\),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+', ' ')


In [25]:
zeroshot_cleaned

ad_creative_body  \
0      it is not enough to defeat donald trump we mus...   
1      we just won a huge victory in the primary elec...   
2      i ve got a small favor to ask can you add your...   
3      the dnc just raised the debate donor threshold...   
4      the enormous student debt crisis is the result...   
...                                                  ...   
57842  nancy pelosi s leading the charge to violate t...   
57843  the consequences of war is that our daughters ...   
57844  nancy pelosi s impeachment hit list keeps grow...   
57845  nancy pelosi s agenda for congress does not re...   
57846  the supreme court just started hearings on the...   

                                            impr_geo_agg  spend_agg  \
0      {'Illinois': 2262508, 'Ohio': 2400000, 'Texas'...   803277.5   
1      {'South Dakota': 40245, 'South Carolina': 1640...   349932.5   
2      {'Alaska': 58937, 'Pennsylvania': 692570, 'Nev...   271561.5   
3      {'Alabama': 139291, 'Pennsylvania': 799612, 'N...   616753.5   
4      {'Alabama': 324699, 'Pennsylvania': 748758, 'N...   221751.5   
...                                                  ...        ...   
57842                                   {'Florida': 500}       49.5   
57843                                     {'Maine': 500}       49.5   
57844                                  {'Nebraska': 500}       49.5   
57845                                     {'Texas': 500}       49.5   
57846                                  {'Delaware': 500}       49.5   

         impr_agg           page_id                 page_name  \
0      25442483.5   124955570892789            Bernie Sanders   
1      24854683.0  1316372698453411  Alexandria Ocasio-Cortez   
2      19366864.5   124955570892789            Bernie Sanders   
3      18443203.5       36872302227               Cory Booker   
4      18399401.5       38471053686          Elizabeth Warren   
...           ...               ...                       ...   
57842       499.5   666196620070127      Maria Elvira Salazar   
57843       499.5   344010372303511                Angus King   
57844       499.5   109592402468562               Deb Fischer   
57845       499.5  1558042497785635            Beth Van Duyne   
57846       499.5        7209432197               Chris Coons   

      ad_creation_time        party  \
0           2020-03-05     Democrat   
1           2020-09-11     Democrat   
2           2020-03-17     Democrat   
3           2019-07-25     Democrat   
4           2020-02-04     Democrat   
...                ...          ...   
57842       2021-10-21   Republican   
57843       2018-08-07  Independent   
57844       2019-11-11   Republican   
57845       2020-04-20   Republican   
57846       2020-04-23     Democrat   

                                                emotions     ANGRY   ANNOYED  \
0      {'AFRAID': 0.09056519009021387, 'AMUSED': 0.16...  0.116733  0.135540   
1      {'AFRAID': 0.09745483939705116, 'AMUSED': 0.13...  0.134677  0.142094   
2      {'AFRAID': 0.07969887694789736, 'AMUSED': 0.15...  0.131562  0.158462   
3      {'AFRAID': 0.09375970552559433, 'AMUSED': 0.13...  0.133796  0.140846   
4      {'AFRAID': 0.10524118111376297, 'AMUSED': 0.12...  0.139728  0.137710   
...                                                  ...       ...       ...   
57842  {'AFRAID': 0.11236538195107602, 'AMUSED': 0.13...  0.168018  0.161798   
57843  {'AFRAID': 0.09867305237088528, 'AMUSED': 0.13...  0.125781  0.135670   
57844  {'AFRAID': 0.0816414466029903, 'AMUSED': 0.140...  0.141876  0.180827   
57845  {'AFRAID': 0.0887159273868603, 'AMUSED': 0.144...  0.129105  0.155833   
57846  {'AFRAID': 0.10437548498299842, 'AMUSED': 0.12...  0.168615  0.157040   

         AMUSED    AFRAID     HAPPY  INSPIRED       SAD  DONT_CARE  \
0      0.166387  0.090565  0.097104  0.184500  0.088420   0.120751   
1      0.130600  0.097455  0.103810  0.171917  0.099821   0.119625   
2      0.151987  0.079699  0.108307  0.165868

In [26]:
merge = pd.merge(zeroshot_cleaned, keywords_cleaned, on='ad_creative_body', how='left')

In [27]:
merge[merge['sample']=='climate_keywords']

ad_creative_body  \
22      unbelievable after stacey abrams helped chang...   
122    named the hill s top climate nerd businessman ...   
181    rep alexandria ocasio cortez and i just introd...   
251    take any of our big problems health care clima...   
329     50 000 responses missing 11 59 p m tonight mi...   
...                                                  ...   
51632  the climate crisis is one of the biggest chall...   
51633  the climate crisis is one of the biggest chall...   
51634  the climate crisis is one of the biggest chall...   
51635  the climate crisis is one of the biggest chall...   
51636  the climate crisis is one of the biggest chall...   

                                            impr_geo_agg  spend_agg  \
22     {'Mississippi': 82272, 'Kansas': 59103, 'Tenne...   261879.5   
122    {'Delaware': 29, 'Oregon': 302, 'Missouri': 11...   187494.0   
181    {'Alabama': 21288, 'Pennsylvania': 162900, 'Ne...    60080.5   
251    {'Vermont': 8757, 'Oregon': 6916, 'Nebraska': ...    44892.5   
329    {'Alaska': 5829, 'Oregon': 55139, 'Minnesota':...   105736.0   
...                                                  ...        ...   
51632                      {'Texas': 2, 'Colorado': 497}       49.5   
51633                      {'Texas': 2, 'Colorado': 497}       49.5   
51634                      {'Texas': 2, 'Colorado': 497}       49.5   
51635                      {'Texas': 2, 'Colorado': 497}       49.5   
51636                      {'Texas': 2, 'Colorado': 497}       49.5   

         impr_agg           page_id                 page_name  \
22     10547980.5   109111900634787  Reverend Raphael Warnock   
122     5362496.0  1934158976830122  Sean Casten for Congress   
181     4325930.5   124955570892789            Bernie Sanders   
251     3509993.0       38471053686          Elizabeth Warren   
329     2879986.0   134498556891640               Val Demings   
...           ...               ...                       ...   
51632       499.5   280704452367167                Jason Crow   
51633       499.5   280704452367167                Jason Crow   
51634       499.5   280704452367167                Jason Crow   
51635       499.5   280704452367167                Jason Crow   
51636       499.5   280704452367167                Jason Crow   

      ad_creation_time     party  \
22          2021-09-23  Democrat   
122         2020-10-22  Democrat   
181         2019-08-07  Democrat   
251         2019-11-12  Democrat   
329         2021-10-27  Democrat   
...                ...       ...   
51632       2021-06-24  Democrat   
51633       2021-06-24  Democrat   
51634       2021-06-24  Democrat   
51635       2021-06-24  Democrat   
51636       2021-06-24  Democrat   

                                                emotions     ANGRY  ...  \
22     {'AFRAID': 0.10363804180351888, 'AMUSED': 0.13...  0.127941  ...   
122    {'AFRAID': 0.09520029463749527, 'AMUSED': 0.15...  0.113530  ...   
181    {'AFRAID': 0.13320879445363867, 'AMUSED': 0.12...  0.146349  ...   
251    {'AFRAID': 0.10195171367028856, 'AMUSED': 0.14...  0.125602  ...   
329    {'AFRAID': 0.11732185281751054, 'AMUSED': 0.13...  0.149745  ...   
...                                                  ...       ...  ...   
51632  {'AFRAID': 0.10799956387941563, 'AMUSED': 0.13...  0.119724  ...   
51633  {'AFRAID': 0.10799956387941563, 'AMUSED': 0.13...  0.119724  ...   
51634  {'AFRAID': 0.10799956387941563, 'AMUSED': 0.13...  0.119724  ...   
51635  {'AFRAID': 0.10799956387941563, 'AMUSED': 0.13...  0.119724  ...   
51636  {'AFRAID': 0.10799956387941563, 'AMUSED': 0.13...  0.119724  ...   

         AMUSED    AFRAID     HAPPY  INSPIRED       SAD  DONT_CARE  \
22     0.134265  0.103638  0.112651  0.166789  0.097768   0.116851   
122    0.155492  0.095200  0.108481  0.174267  0.084025   0.142153   
181    0.121521  0.133209  0.098221  0.153596  0.097866   0.104601   
251    0.143505  0.101952  0.097674  0.173508  0.101257   0.116497   
3

In [28]:
merge_view = merge[['ad_creative_body', 'zero_shot_label', 'sample']]

In [29]:
merge_view.zero_shot_label.unique()

#zeroshot labels related to climate change:
climate_zeroshot_labels = ['climate', 'education', 'environment',
       'climate, economy', 'climate, immigration',
       'environment, education', 'climate, education',
       'climate, inequality',
       'environment, economy']

In [30]:
merge_view[merge_view['zero_shot_label'].isin(climate_zeroshot_labels) == True]

ad_creative_body zero_shot_label  \
22      unbelievable after stacey abrams helped chang...         climate   
122    named the hill s top climate nerd businessman ...         climate   
163    rick scott subtracted 1 3 billion from public ...       education   
181    rep alexandria ocasio cortez and i just introd...         climate   
203     when it comes to progressive leadership it s ...       education   
...                                                  ...             ...   
51632  the climate crisis is one of the biggest chall...         climate   
51633  the climate crisis is one of the biggest chall...         climate   
51634  the climate crisis is one of the biggest chall...         climate   
51635  the climate crisis is one of the biggest chall...         climate   
51636  the climate crisis is one of the biggest chall...         climate   

                 sample  
22     climate_keywords  
122    climate_keywords  
163                 NaN  
181    climate_keywords  
203                 NaN  
...                 ...  
51632  climate_keywords  
51633  climate_keywords  
51634  climate_keywords  
51635  climate_keywords  
51636  climate_keywords  

[1827 rows x 3 columns]

zeroshot finds 1827 unique ad_creative_body texts with the climate labels applied

keyword finds 1339 unique ad_creative_body texts with related to climate change

527 of the 1827 zeroshot unique ad_creative_body texts are not found by the keywords approach

1330 of the 1339 keyword unique ad_creative_body texts are also in the zeroshot sample.
